### ISSUE TITLE
[Feature Request] SpriteBatch

### ISSUE THREAD

**Issue by RUSshy** [mellinoe/veldrid#79]
Hello

Is it possible to add a SpriteBatch class similar to MonoGame ? So it makes easier to switch to this library

Thanks

**Comment by mellinoe** [comment]
It probably wouldn't go in the core Veldrid library (too high level), but I'm not opposed to having an "extension" library with commonly-requested stuff like this in it. At the very least, it would be good to have a project in the [samples repo](https://github.com/mellinoe/veldrid-samples) with code to render sprites.

I'm pretty busy with other work at the moment, but I would like to have this at some point.

**Comment by rcollina** [comment]
Hi @mellinoe, 
Thanks for Veldrid!
I'd love to see some 2D tutorials. I admit I don't have much graphics programming experience, and what little I had is lost in time. :) 
Still, I think it would be a boon for adoption, as many indie devs focus on 2d-oriented games. My two cents. Thanks again!

**Comment by LithiumToast** [comment]
I would be interested in keeping a mirror of [MonoGame.Extended](https://github.com/craftworkgames/MonoGame.Extended) for Veldrid if there is enough need for it. Full disclosure I'm a contributor. `SpriteBatch` could be one thing added quite easily in such a library.

**Comment by rcollina** [comment]
@LithiumToast that would be awesome.

**Comment by LithiumToast** [comment]
As for the content pipeline in MonoGame I was also thinking of creating a replacement; I'm calling it [Assetyard](https://github.com/Styxz/Assetyard)

**Comment by rcollina** [comment]
It'd be cool to either have a cli (something like "dotnet veldrid-assets build", a global tool) or an automatic build when checksums change. (https://natemcmaster.com/blog/2017/11/11/build-tools-in-nuget/)

**Comment by vpenades** [comment]
@LithiumToast @mellinoe If you're considering some kind of Content Asset Processor like MonoGame, I would like you to pay attention to one of my projects:

https://github.com/vpenades/UberFactory

It's a general purpose content pipeline. It's plugin based, so you can theoretically write plugins to deliver content for any game engine/platform.

**Comment by LithiumToast** [comment]
@vpenades Very neat.

**Comment by rcollina** [comment]
@vpenades looks great! I gave the repo a cursory look. 
Will it be possibile to deliver UberFactory as a [global tool](https://surfingthecode.com/fun-with-global-tools-in-.net-core-2.1-preview/) at some point? I believe global tools have a great potential.
Can you point me to some documentation so I can understand how you structured UberFactory?
Thanks!

**Comment by RUSshy** [comment]
Omg please no, i hope there is no forced content pipeline like monogame.. 
If u want that please do something like https://github.com/tlgkccampbell/ultraviolet

You can load everything from file (texture, audio, shared, sprite, and even shaders)

Like seriously, who need to add bloat for just loading BitmapFont / Textures that are just few kbs, not everyone will make AAAAAAAAAA games..

**Comment by LithiumToast** [comment]
@RUSshy For development Content Pipeline can be a pain yes. For deployment purposes please go to this [page](https://docs.microsoft.com/en-ca/xamarin/graphics-games/cocossharp/content-pipeline/introduction#why-use-a-content-pipeline) and read it.

**Comment by rcollina** [comment]
I do agree there’s value in keeping things simple, however, as long as the content pipeline is automated, not much is going to actually change for a developer.

**Comment by feliwir** [comment]
The content pipeline shouldn’t be part of core veldrid in my opinion. I have no issues with a seperate NuGet though.

**Comment by LithiumToast** [comment]
@feliwir Could I use your code as inspiration? I'll comply with the GNU Lesser obviously. Even link to your repository from mine.

**Comment by vpenades** [comment]
@rcollina yes, the idea is that the CLI tool can be used as a global tool. For basic usage, I wrote this: https://github.com/vpenades/UberFactory/issues/1 but it might be a bit outdated.

@RUSshy It's just about having multiple choices... the point with content pipeline managers is, when you target multiple platforms, assets need to be tailored per platform, and even for small projects, it becomes a big management problem.

Anyway, I didn't want to steal this thread, if you're interested in uber factory, feel free to post in the repo's issues page.

@feliwir I agree... one of the pains of Monogame is that its goods and bads, its content pipeline is its only choice.

**Comment by LithiumToast** [comment]
@vpenades Could I use your code as inspiration? I'll comply with the GNU Lesser obviously. Even link to your repository from mine.

**Comment by mellinoe** [comment]
There are a couple of implementations out there, but I haven't tested one extensively. It is still on my list of sample projects to create.

@Jjagg Would you recommend OpenWheels to someone looking for a Veldrid-backed sprite renderer? I think you have the most complete implementation that I have seen.

**Comment by Jjagg** [comment]
I'd like to recommend it, but it's a bit early for that. I've only used it in some toy projects, just getting started on a serious project with it so I don't want to lock down the API yet. I've already considered some major design changes. I welcome feedback from anyone willing to try it though. That helps getting the API right.

@Thraka I'm curious for your motivation to move away from MonoGame.
A good solution for a SpriteBatch replacement might be to fork OpenWheels and make changes as required. Please do post any feedback in the form of an issue.
Here's a link to the GitHub page: https://github.com/Jjagg/OpenWheels 
The readme is bleak, but the API is pretty well-documented. If you want to get started with the Batcher class take a look at the documentation in the [source code](https://github.com/Jjagg/OpenWheels/blob/462321522aa6ec799782b25d9888427426c011df/src/OpenWheels.Rendering/Batcher.cs#L8-L38). I haven't setup a web page for the docs yet.

**Comment by Thraka** [comment]
@Jjagg Cool, I'll check it out!

I have a few reasons for investigating moving away from MonoGame (or supporting something else in addition to MonoGame. I did had support for SFML at one point but dropped it)

1. I want to get out of framework and move to .NET Core. While there is the unofficial template does work, I've had oddities with it where it's running slow one minute and fast another.

2. I've been trying to integrate a lot more with https://github.com/Chris3606/GoRogue/ and since it's a generic lib, it uses its own primitive types, so converting between types is annoying and confusing for my users. Having a framework where I can specify my own primitive type (or in this case, use GoRogue's) eases that user confusion

3. I do have some odddd issues with MonoGame across multiple computers over the last few years. (I filed a bug/forum post about it a while back, but I can't find it. Maybe I used a diff account) basically what happens is at some point my system stops responding well to MonoGame. Release, Debug, in VS, outside of VS, I've never been able to track it down to something. As soon as it starts though, it doesn't go away until I reboot. When this issue appear (driver? something? no idea, no other games have a problem), any mouse movement over the window of MonoGame causes the framerate to become crippled. Stop moving the mouse, it goes back to full speed. Closing the window and restarting the game does not change anything. Only rebooting the computer seems to fix it. Then it doesn't appear for 2 or so days, then it's back and there until I reboot. It's very odd.

4. It's also fun to start over and play with something new

**Comment by Jjagg** [comment]
Neat :) The spacing for the font rendering looks terrible though.

The exception occurs because the Veldrid backend for OpenWheels doesn't grow the vertex and index buffers yet. It leaves them at the default size. Should definitely get that fixed.

**Comment by Thraka** [comment]
@Jjagg Spacing is working as it should, considering this is a pure spritesheet font, which makes it monospaced. It's based on the IBM font from MSDOS.

**Comment by shihan42** [comment]
Hello @mellinoe @Thraka,

any updates on this one? Especially for a GUI, a sprite batch would be nice and useful.
I know there is ImGui, which I absolutely love, but that's only suitable for debug and/or development, but not for a finished and polished end-user ingame GUI :)

**Comment by Thraka** [comment]
I wrote a spritebatch like class for SFML using some examples and piecing things together.  If you have to put something of your own together, you could use this as a basis to create something.

https://gist.github.com/Thraka/9460b6988f8eb629dbfb01e3770602b1

**Comment by lithiumtoast** [comment]
I'm writing a `SpriteBatch` for `sokol` this week/weekend (OpenGL Core 3.3); I could drop by with a version for Veldrid if you guys are interested.

**Comment by feliwir** [comment]
I thought i already mentioned this, but we have a Veldrid SpriteBatch class in [OpenSage](https://github.com/OpenSAGE/OpenSAGE), which you can find [here](https://github.com/OpenSAGE/OpenSAGE/blob/master/src/OpenSage.Game/Graphics/SpriteBatch.cs). Of course it's not a standalone solution, but you might find some useful concepts there

**Comment by shihan42** [comment]
Thanks for your comments. There is some free time this weekend that I'll use to check out your concepts.

@lithiumtoast That would be nice!

**Comment by lithiumtoast** [comment]
I have a very basic approach working in OpenGL 3.3 but it's not Veldrid; I will report back later with a Veldrid version when I finish experimenting with a different version for modern graphics APIs in the next week or two. The basic version is not very complicated. I use the classic vertex declaration found in XNA/MonoGame with colour packed as RGBA bytes.

```cs
private struct Vertex
{
    public Vector3 Position;
    public Vector2 TextureCoordinates;
    public uint Color;
}
```

Things I want to try with an experimental version with modern graphics API:

1. Try instancing with a `Matrix3x2` per quad to transform the model vertices (defined in the vertex shader) and drop the position data from the vertex declaration.
2. Try instancing with a `Vector4` or `Rectangle` per quad to translate and scale the model texture coordinates (defined in the vertex shader) and drop the texture coordinates from the vertex declaration.
